monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/AngleMonitoring.java

```diff
@@ -0,0 +1,373 @@
+                networkActionOk = networkActionOk && checkElementaryActionAndStoreInjection(ea, networkClone, angleCnecId, na.getId(), networkElementsToBeExcluded, powerToBeRedispatched);
+                if (!networkActionOk) {
+                    break;
+                }
+            }
+            if (networkActionOk) {
+                na.apply(networkClone);
+                appliedNetworkActions.add(na);
+            }
+        }
+        return appliedNetworkActions;
+    }
+
+    /**
+     * 1) Checks a network action's elementary action : it must be a Generator or a Load injection setpoint,
+     * with a defined country.
+     * 2) Stores applied injections on network
+     * Returns false if network action must be filtered.
+     */
+    private boolean checkElementaryActionAndStoreInjection(ElementaryAction ea, Network networkClone, String angleCnecId, String naId, Set<String> networkElementsToBeExcluded, Map<Country, Double> powerToBeRedispatched) {
+        Map<Country, Double> temporaryPowerToBeRedispatched = powerToBeRedispatched;
+        if (!(ea instanceof InjectionSetpoint)) {
+            BUSINESS_WARNS.warn("Remedial action {} of AngleCnec {} is ignored : it has an elementary action that's not an injection setpoint.", naId, angleCnecId);
+            return false;
+        }
+        // Elementary actions are either generators or loads
+        Identifiable<?> ne = networkClone.getIdentifiable(((InjectionSetpoint) ea).getNetworkElement().getId());
+        Optional<Substation> substation = ((Injection<?>) ne).getTerminal().getVoltageLevel().getSubstation();
+        if (substation.isEmpty()) {
+            BUSINESS_WARNS.warn("Remedial action {} of AngleCnec {} is ignored : it has an elementary action that doesn't have a substation.", naId, angleCnecId);
+            return false;
+        } else {
+            Optional<Country> country = substation.get().getCountry();
+            if (country.isEmpty()) {
+                BUSINESS_WARNS.warn("Remedial action {} of AngleCnec {} is ignored : it has an elementary action that doesn't have a country.", naId, angleCnecId);
+                return false;
+            } else {
+                checkGlsks(country.get(), naId, angleCnecId);
+                if (ne instanceof Generator) {
+                    temporaryPowerToBeRedispatched.merge(country.get(), ((Generator) ne).getTargetP(), Double::sum);
+                } else if (ne instanceof Load) {
+                    temporaryPowerToBeRedispatched.merge(country.get(), -((Load) ne).getP0(), Double::sum);
+                } else {
+                    BUSINESS_WARNS.warn("Remedial action {} of AngleCnec {} is ignored : it has an injection setpoint that's neither a generator nor a load.", naId, angleCnecId);
+                    return false;
+                }
+                networkElementsToBeExcluded.add(ne.getId());
+            }
+        }
+        powerToBeRedispatched.putAll(temporaryPowerToBeRedispatched);
```

Comment:
You still have the issue that powerToBeRedispatched is updated if the first elementaryAction is valid and not the second elementaryAction (and the networkAction not applied). 
One solution would be to have one method just check the validity of the elementary actions, and one method just store the injection. Then in your for network action loop, in the for elementary action loop, you check the validity and update the networkActionOk boolean as you go, and if it's still true at the end of that for loop, you loop on the elementary actions once more to add the injections. 
Another solution is to also create an empty map when you create the networkActionOk boolean, and call this method with that map instead of powerToBeRedispatched. Then when you actually apply the network action (in the if), you merge powerToBeRedispatched with the map you've created (and that this method will have completed).